File: frontend/src/api/goals.ts
import apiClient from './client';
import {
  Goal,
  GoalListResponse,
  CreateGoalRequest,
  UpdateGoalRequest,
} from '../types';

export const goalsAPI = {
  list: async (activeOnly: boolean = true): Promise<GoalListResponse> => {
    const response = await apiClient.get<GoalListResponse>('/goals', {
      params: { active_only: activeOnly },
    });
    return response.data;
  },

  create: async (data: CreateGoalRequest): Promise<Goal> => {
    const response = await apiClient.post<Goal>('/goals', data);
    return response.data;
  },

  update: async (goalId: string, data: UpdateGoalRequest): Promise<Goal> => {
    const response = await apiClient.patch<Goal>(`/goals/${goalId}`, data);
    return response.data;
  },

  delete: async (goalId: string): Promise<void> => {
    await apiClient.delete(`/goals/${goalId}`);
  },
};

File: backend/app/routes/goals.py
from uuid import UUID

from fastapi import APIRouter, Depends, Query
from sqlalchemy.orm import Session

from app.core.dependencies import get_current_user
from app.db.database import get_db
from app.models.user import User
from app.schemas.goal import GoalCreate, GoalUpdate, GoalResponse, GoalListResponse
from app.services.goal_service import GoalService

router = APIRouter()


@router.post("", response_model=GoalResponse, status_code=201)
async def create_goal(
    data: GoalCreate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    return GoalService.create(db, data, current_user)


@router.get("", response_model=GoalListResponse)
async def list_goals(
    active_only: bool = Query(True),
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    return GoalService.list(db, current_user, active_only=active_only)


@router.patch("/{goal_id}", response_model=GoalResponse)
async def update_goal(
    goal_id: UUID,
    data: GoalUpdate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    return GoalService.update(db, goal_id, data, current_user)


@router.delete("/{goal_id}")
async def delete_goal(
    goal_id: UUID,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    return GoalService.delete(db, goal_id, current_user)
